### ISSUE TITLE
Memory leaking with large object serialization

### ISSUE THREAD

**Issue by TylerBrinks** [JamesNK/Newtonsoft.Json#2671]
### Expected behavior

Serializing or deserializing types should recover allocated memory.  Unmanaged memory should remain at a relatively consistent size after GC has finished executing, Gen1 and LOH in particular.

### Actual behavior

In the attached sample Web API project, a model object with a large dictionary is constructed and serialized JSON as the response.  Repeat requests to the endpoint cause memory to increase without any upper bound.  GC eventually runs and handles objects via Gen0 and LOH/POH (seemingly little fragmentation) but the overall unmanaged memory footprint continues to grow on each request without any limit.  I've been able to get the memory consumption to over 10GB.

### Steps to reproduce

Run the attached project.  The default API endpoint's URL should open in your default browser.  Refresh the page repeatedly and observe memory usage.  I have set the launch settings to run the app in a console to make it as simple as possible to profile.

I've included a custom JsonConverter which speeds up the problem, however the same symptoms exist with or without the use of the JsonConverter.  

Profiling the memory usage shows the overall unmanaged size growing steadily with each new request.
![Screenshot 2022-04-23 141501](https://user-images.githubusercontent.com/121930/164945329-9d8e2a9c-fa22-42da-9530-b54abd1fa4f3.png)

At first glance, the number of JObects seems to be hanging around between requests, even after GC completes.  After only a few iterations, over 3 million instances.
![Screenshot 2022-04-23 142242](https://user-images.githubusercontent.com/121930/164945439-c7823dc2-3159-47ca-a728-9ba0b235de93.png)

It appears most of the execution effort is spent rendering said JToken objects.
![Screenshot 2022-04-23 142340](https://user-images.githubusercontent.com/121930/164945473-309fd197-9cfd-4574-b835-941a894da3a4.png)

With my particular profiling session and just a handful of requests, I reached over 10 million JObjects with 1GB of consumed memory.  The objects, values, and dictionary combined represent around 35 million objects.  Compare this to the 3 million JObjects in the screenshot above which, after subsequent HTTP requests and GC cycle still shows the object total more than tripling.
![Screenshot 2022-04-23 142822](https://user-images.githubusercontent.com/121930/164945543-20e228c7-3556-45f6-839d-cebd66f03f3d.png)

Final result during a typical profiling session shows the high consumption of this particular method chain.
![Screenshot 2022-04-23 142615](https://user-images.githubusercontent.com/121930/164945769-4b947965-a257-4d5b-93d3-7493ca4992b4.png)

Attached sample project.  Unzip, run, refresh, repeat.
[WebApplication1.zip](https://github.com/JamesNK/Newtonsoft.Json/files/8548589/WebApplication1.zip)

**Comment by sungam3r** [comment]
If this is really a memory leak and not just its large consumption with the subsequent return to the system, then some objects should remain in memory (JObject?). Could you show ones and provide retention graph for them from dotMemory?
Another thought regarding - memory heap fragmentation after many huge allocation. If I remember correctly dotMemory shows fragmentation statistics for memory snapshots.

**Comment by JamesNK** [comment]
I can see a saw-tooth pattern in your memory trace. That's normal and fine. See https://plumbr.io/blog/memory-leaks/memory-leaks-fallacies-and-misconceptions. It's about Java but applicable to .NET.

If the 10 million objects are collected when GC runs then there isn't a memory leak.

**Comment by prince-melvin** [comment]
Hi, Any luck on this issue?

I am facing the same issue when deserializing large object. (JSON.Parse())
From my endpoint I am returning the response as a JObject. It is mentioned here as unused memory but it is not claimed by the GC.

And upon multiple Force GC's like 10-12 times it is getting reduced.

![snapshot_mem](https://user-images.githubusercontent.com/81223599/184835651-cc29f422-8eb7-49c8-8fcb-fdb03bf3f17f.PNG)

**Comment by TylerBrinks** [comment]
@prince-melvin No, I've had to resort to an alternative.

**Comment by justintien** [comment]
I had the same problem.

When I
```c#
JToken.Parse(jsonString); // big json (e.g. 30MB)
```

It will take up 900MB of memory, and can't be release the memory (even I have no reference it)

And I tested if data use protobuf, memory can be reduced by a factor of 1-2, and when no reference, the memory will be release.

So about JToken.Parse I missing what options?

**Comment by tekguy** [comment]
> @prince-melvin No, I've had to resort to an alternative.

@TylerBrinks what other solutions you came up with? Running into the same issue with my API very similar to your POC, my data is coming from redis to deserialize

**Comment by TylerBrinks** [comment]
> > @prince-melvin No, I've had to resort to an alternative.
> 
> @TylerBrinks what other solutions did you come up with? Running into the same issue with my API very similar to your POC, my data is coming from redis to deserialize.

No, I had to switch to the .net 7+ System.Text option instead.

**Comment by ChrisBertrand** [comment]
@TylerBrinks Did the switch to System.Text.Json solve the problem for you?

Do we have a workaround for this, or a proposed solution? @JamesNK Is this something you could help with, or do you not foresee this as a problem at the moment?

**Comment by TylerBrinks** [comment]
> @TylerBrinks Did the switch to System.Text.Json solve the problem for you?
> 
> Do we have a workaround for this, or a proposed solution? @JamesNK Is this something you could help with, or do you not foresee this as a problem at the moment?

Yes, the problem went away after switching to the System.Text.Json serializer

**Comment by patrickkulling** [comment]
Was there any development in that direction on reducing/freeing up the heap memory again?
We can confirm that we do see the same issue in our project.

**Comment by TylerBrinks** [comment]
> Was there any development in that direction on reducing/freeing up the heap memory again? We can confirm that we do see the same issue in our project.

No. I had to switch to System.Text.Json in my case.

**Comment by mu88** [comment]
@TylerBrinks / @JamesNK : I'm facing a similar issue: I'm deserializing thousands of smaller JSON strings (about 300 KB each)  coming from EF Core, leaving me with 30 MB managed and 800 MB unmanaged memory - and some `OutOfMemory` exceptions.

The code looks more or less like this:
```csharp
public class MyContext : DbContext
{
    public DbSet<Point> Points { get; set; }
}

public class Point
{
    public string Key { get; set; }

    public string Content { get; set; }

    public PointDto GetObjectFromContent() => JsonConvert.DeserializeObject<PointDto>(Content);
}

public record class PointDto(string Name, string Address);

public enum State
{
    Received,
    Processed
}

public class MyService(IDbContextFactory<MyContext> contextFactory)
{
    public async Task<List<string>> GetNamesAsync()
    {
        using MyContext context = contextFactory.CreateDbContext();
        return await context.Points
            .Where(point => point.State == State.Received)
            .Select(point => point.GetObjectFromContent().Name)
            .ToListAsync();
    }
}
```

Here's the exception:
```
System.OutOfMemoryException: Exception of type 'System.OutOfMemoryException' was thrown.
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.PopulateObject(Object newObject, JsonReader reader, JsonObjectContract contract, JsonProperty member, String id)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
   at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
   at Newtonsoft.Json.JsonConvert.DeserializeObject(String value, Type type, JsonSerializerSettings settings)
```

I'm curious about how `System.Text.Json` helps here - thx for your time 🤜🏻🤛🏻

**Comment by TylerBrinks** [comment]
I had to switch to System.Text.Json.  I have no memory issues after I made the swap.

**Comment by mu88** [comment]
Unfortunately, that's not feasible in my case 🫤 

Let's see whether @JamesNK has an idea about what might be going wrong in my case. Furthermore, I'd be curious to know what's going on here under the hood: I'm struggling to understand why deserializing a bunch of strings can lead to such a mass of unmanaged memory.

**Comment by JamesNK** [comment]
No one here has shown there is a memory leak in Newtonsoft.Json. If you provide a memory trace that shows Newtonsoft.Json rooting unmanaged memory then I'll take a look at it.

Until I see a memory leak, I think it's far more likely that you're either serializing/deserializing a large amount of JSON on a machine with a memory limit (S.T.J would be worth investigating because it's more efficent with memory), or there is a leak somewhere else in your application.

**Comment by mu88** [comment]
Oh, I didn't want to give the impression that I believe that Json.NET is leaking memory. Sorry if you got this feeling 🫤

I was just searching around for an existing issue that fits my observation of "when deserializing a lot of JSON strings, the unmanaged memory explodes" to get some help - and this issue seemed to fit nicely.
